### ISSUE TITLE
Monitoring incoming traffic

### ISSUE THREAD

**Issue by ghrlt** [httptoolkit/httptoolkit#284]
Hi! First, thanks for having created such an awesome tool!
I'm facing a problem, where I would need to monitor **incoming** traffic going to my android/chrome/anything..

As of today, is there a way to do it? If so, how?
If not, could it be a futur enhancement? 

---

_Does this affect you too? Click the +:smiley: icon in the top right and add a :+1: on issues to help decide where HTTP Toolkit goes next, or [go take a look & vote on the most popular ideas so far](https://github.com/httptoolkit/feedback/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc)._

**Comment by shirshak55** [comment]
Sorry, but can't you select android? In android, you can download HTTP toolkit? For some, you can use proxy extension (but you might need to install certificates). Some apps might pin certs so you might have to reverse it.

**Comment by ghrlt** [comment]
Of course I can. I'm obviously able to use HTTP Toolkit to monitor **outgoing** HTTP(s) requests, but not for **incoming** HTTP(s) requests..

**Comment by pimterry** [comment]
This is possible, but I think only with HTTP Toolkit Pro, or funky client configuration.

Basically you add a forwarding rule in HTTP Toolkit that says "send all requests to real server URL X", and then you tell every client to connect to HTTP Toolkit directly instead of that real server URL. HTTP Toolkit will receive direct requests, and it'll show them all, and then forward them to the real server so they work as normal. That rule would look like this:

![Screenshot from 2022-05-27 17-14-11](https://user-images.githubusercontent.com/1526883/170728437-bf019c18-d601-4f34-8846-874aa880f064.png)

You can test this without Pro (by setting breakpoints, and manually editing the URL of every incoming request) but it'll be much more inconvenient than just adding a general redirect rule.

Bear in mind you'll also need to either configure the clients to use a plain HTTP URL, or to trust HTTP Toolkit's CA certificate though.

Does that make sense, and does that work for your use case? Purely out of curiosity, can you explain more about what you're trying to do?

**Comment by pimterry** [comment]
No, the above is a way to listen to incoming requests to the monitored device. Somehow you need to change the clients to talk to HTTP Toolkit instead of talking directly to the device though.

Basically you have this:

```
Clients --> monitored device
```

You need to get to this:

```
Clients --> HTTP Toolkit --> monitored device
```

There's no way HTTP Toolkit can show traffic that never goes to HTTP Toolkit somehow.

To do that, you need to make both of the arrows there work: you need to redirect traffic from the clients to HTTP Toolkit (the left arrow), and you need to ensure the traffic afterwards is sent to the monitored device (the right arrow).

The redirection rule above handles the right arrow, by sending all received requests to the monitored device address automatically. Hopefully that makes sense.

For the left arrow, the 'best' option is always going to be changing the client config to just use HTTP Toolkit's address instead of the monitored device's address. That's what I was assuming above. This doesn't mean using HTTP Toolkit as a proxy - they can connect directly as if it was a server, and HTTP Toolkit will handle that - but they would need some kind of reconfiguration.

If that isn't possible, the options depend on your setup. For example, you could change the domain name used for your monitored device to point to HTTP Toolkit instead (and change the ports accordingly) or you could use iptables or some other firewall on your network to redirect TCP packets for the monitored device through HTTP Toolkit instead.

That part really depends on your setup and what you wan to do, so it's hard to help without more context. Somehow you have to make the requests from the clients go to HTTP Toolkit instead though (and make them trust HTTP Toolkit, if you're using HTTPS). There is no possible world where HTTP Toolkit can show this data without somehow getting in between the clients and your monitored device.

**Comment by ghrlt** [comment]
Oh ! I understand now, thanks.

However, I'm not sure how I could implement this for an Android device.
Let's suppose that I have a third-party app, and that I want to listen to all the requests this 3rd party app's server is making to its app, on my phone. Should I implement a firewall on my phone to redirect all requests to http toolkit? And how would the 3rd party app's server trust HTTP Toolkit certificate, as I'm 100% using https?

**Comment by pimterry** [comment]
It's very very unusual for a server to make requests to an Android app. Are you sure that's what's happening? They're not doing two-way communication over websockets or gRPC or something, where the connection is initiated by the Android device?

It's unusual because it won't work in many mobile environments at all. NAT on home networks and probably many mobile networks will block all incoming connections to local devices. I'd be very surprised if any popular app did this, because it just won't work for most people.

I'm also not sure how HTTPS could work here. HTTPS generally works using certificates linked to domain names, so you'd need a domain name that resolved to the address of your Android device (which is not fixed, so that would need constant updating). It could technically be possible without that using self-signed certificates and certificate pinning, but it would be very very odd.

Are you _sure_ that's what's happening? Why?

If that really is happening, then it's technically possible to do this kind of redirection using Android's VPN APIs to rewrite packets (HTTP Toolkit does this for outgoing requests already, more details here: https://httptoolkit.tech/blog/inspecting-android-http/). Alternatively you could do so on the local network level, by redirecting all network traffic for your phone via your computer, e.g. with ARP spoofing or something, and then using IPTables or similar on your computer to rewrite the packets before you forward them to your phone. Both options will be complicated!

The HTTPS part would be a bit easier though - in any plausible case like that, the HTTPS private key & cert must live in the Android app on your phone, so you can probably extract it from the APK, and then you can just use the real app's certificate directly.

**Comment by ghrlt** [comment]
Hmm, I'm pretty sure you're right. I'm sorry for my misunderstanding, and thanks for your explanation

Then, is it possible to monitor websockets connections? Or even mqtt traffic?

**Comment by pimterry** [comment]
MQTT definitely not I'm afraid, it's a very different protocol, and honestly I'm not too sure about the details of how that works. I think it's sufficiently separate from HTTP (in both the details and the use cases) that it wouldn't make much sense here.

WebSockets meanwhile aren't supported today, but they definitely will be supported, very very soon (see #36). Watch this space! You can subscribe to that issue to get GitHub notifications when it's available if you like.

**Comment by pimterry** [comment]
WebSockets have since been added (in #36) and MQTT is tracked separately in #254 so I'm going to close this.
